### PR TITLE
 fix #526

### LINKED ISSUE #526: The loss CatagoricalCrossEntropy is currently unusable in framework
The use of  class org.tensorflow.framework.losses.CatagoricalCrossEntropy.java ended systematically with this issue :

Exception in thread "main" org.tensorflow.exceptions.TFInvalidArgumentException: Shape must be rank 1 but is rank 0 for '{{node SoftmaxCrossEntropyWithLogits/Slice}} = Slice[Index=DT_INT64, T=DT_INT64](SoftmaxCrossEntropyWithLogits/Shape, SoftmaxCrossEntropyWithLogits/Sub, SoftmaxCrossEntropyWithLogits/Const)' with input shapes: [2], [], [].
	at org.tensorflow.internal.c_api.AbstractTF_Status.throwExceptionIfNotOK(AbstractTF_Status.java:87)
	at org.tensorflow.GraphOperationBuilder.finish(GraphOperationBuilder.java:461)
	at org.tensorflow.GraphOperationBuilder.build(GraphOperationBuilder.java:100)
	at org.tensorflow.GraphOperationBuilder.build(GraphOperationBuilder.java:71)
	at org.tensorflow.op.core.Slice.create(Slice.java:90)
	at org.tensorflow.framework.op.nn.SoftmaxCrossEntropyWithLogits.flattenOuterDims(SoftmaxCrossEntropyWithLogits.java:184)
	at org.tensorflow.framework.op.nn.SoftmaxCrossEntropyWithLogits.softmaxCrossEntropyWithLogits(SoftmaxCrossEntropyWithLogits.java:116)
	at org.tensorflow.framework.op.NnOps.softmaxCrossEntropyWithLogits(NnOps.java:145)
	at org.tensorflow.framework.losses.Losses.categoricalCrossentropy(Losses.java:253)
	at org.tensorflow.framework.losses.CategoricalCrossentropy.call(CategoricalCrossentropy.java:256)
	at org.tensorflow.framework.losses.impl.AbstractLoss.call(AbstractLoss.java:69)
	at org.genericsystem.keras.losses.SparseCategoricalCrossentropyLoss.internal(SparseCategoricalCrossentropyLoss.java:61)
	at org.genericsystem.keras.losses.NativeLossLayer.internalCall(NativeLossLayer.java:23)
	at org.genericsystem.keras.layers.Layer.call(Layer.java:31)
	at org.genericsystem.keras.layers.Layer.call(Layer.java:23)
	at org.genericsystem.keras.model.ModelContext.initalize(ModelContext.java:59)
	at org.genericsystem.keras.model.TrainingContext.initalize(TrainingContext.java:28)
	at org.genericsystem.keras.model.Model.trainStep(Model.java:260)
	at org.genericsystem.keras.example.gan.AbstractACGanApp$3.trainStep(AbstractACGanApp.java:67)
	at org.genericsystem.keras.model.Model.train(Model.java:453)
	at org.genericsystem.keras.model.Model.fit(Model.java:433)
	at org.genericsystem.keras.example.gan.AbstractGanApp.lambda$3(AbstractGanApp.java:117)
	at org.genericsystem.keras.GSKeras.safeExecuteSession(GSKeras.java:164)
	at org.genericsystem.keras.example.gan.AbstractGanApp.fit(AbstractGanApp.java:109)
	at org.genericsystem.keras.example.gan.acgan.cifar10.ACGanApp.main(ACGanApp.java:26)
	
	This is due to the method flattenOuterDims in org.tensorflow.framework.op.nn.SoftmaxCrossEntropyWithLogits.java which build attribute rankminusone as a scalar instead of an array (rank1)

### PR DESCRIPTION
**Pull request by nfeybesse**

### CODE REVIEW

**Comment by Craigacp** [comment]
Can you add a test which triggers this bug to the cross entropy tests - https://github.com/tensorflow/java/blob/master/tensorflow-framework/src/test/java/org/tensorflow/framework/losses/CategoricalCrossentropyTest.java. I think this used to work so I worry it's due to a TF upgrade and we didn't catch it with tests.

**Comment by nfeybesse** [comment]
No, the problem is older, and it is probably the dynamic batch size which triggers the problems. I will try to do a test case

**Comment by nfeybesse** [comment]
@Test
	public void testCategoricalCrossEntopyWithDynamicBatchSize() {
		try (Graph graph = new Graph()) {
			Ops tf = Ops.create(graph);
			Operand<TFloat32> yPred = tf.placeholder(TFloat32.class, Placeholder.shape(Shape.of(-1, 3)));
			Operand<TFloat32> yTrue = tf.reshape(tf.constant(new float[] { 1f, 0f, 0f, 0f, 1f, 0f, 0f, 0f, 1f }), tf.array(3, 3));
			CategoricalCrossentropy instance = new CategoricalCrossentropy(true);
			Operand<TFloat32> loss = instance.call(tf, yTrue, yPred);// Throw TFInvalidArgument Exception without fix
			try (Session session = new Session(graph); TFloat32 result = (TFloat32) session.runner().feed(yPred, TFloat32.tensorOf(Shape.of(3, 3), DataBuffers.of(new float[] { 1.f, 0.f, 0.f, 0.f, 1.f, 0.f, 0.f, 0.f, 1.f }))).fetch(loss).run().get(0)) {
				if (Math.abs(0.5514477f - result.getFloat()) > 0.01)
					throw new IllegalStateException("Invalid result :" + result.getFloat());
			}
		}
	}

**Comment by nfeybesse** [comment]
I am confused because I have the impression that there has not yet been a test carried out by feeding a model with batches of dynamic size. I know from experience that it is largely possible, but that you have to track down a few small bugs. How would you integrate my test so that it would be suitable?

**Comment by Craigacp** [comment]
Add it next to the other tests for that loss. If there are more issues then let's fix them.

More of the framework was in flight a couple of years ago, but we didn't get all of it merged, so I assume that some of those things were tested in the original codebase before it was broken up into smaller PRs.

**Review by Craigacp — APPROVED** [review]

**Comment by Craigacp** [comment]
Thanks
